### PR TITLE
only query for db-metadata once per sync operation + scan each table …

### PR DESCRIPTION
**Pull request by escherize**
# Manual Backport of #26054 for 45 release

* pass in db-metadata where it is used,
- rather than querying for it in random functions
* update callsite of sync-tables-and-database!
* let sync-metabase-metadata! look up db-metadata
- only when needed
* pass catalog (db), and schema to .getPrimaryKeys
* let describe-table* add in the nil table-name
* cleanup
* use :snowflake instead of driver for inheritance
* concurrent-sync-test should call describe-database once
* cleaning up
* call add-table-pks with 3 args
* call add-table-pks with 3 args
* try escaping entity names to appease Oracle Driver
- Oracle wants the table-name to be escaped
- escape is usually a no-op
* pass driver to escape entity
* pull get-table-pks into a multimethod,
- handle the discrepancy directly in Oracle driver
* remove inline def
* fix colliding consistent-namespace linting
- metabase.driver.sql-jdbc.common and metabase.driver.sql-jdbc.sync.common were mapped to sql-jdbc.common
- metabase.driver.sql-jdbc.sync.common is now mapped to sql-jdbc.sync.common
* fix reflection warnings
* apply fix for getting oracle pks
* nix an unused arg

### CODE REVIEW

File: modules/drivers/oracle/src/metabase/driver/oracle.clj
Comment (camsaul) [inline modules/drivers/oracle/src/metabase/driver/oracle.clj:452]:
```suggestion
  [_driver ^Connection conn _db-name-or-nil table]
```

It's a little easier to read this way. Otherwise you have to go look at the `defmulti` to figure out which parameter we're not using.

File: modules/drivers/oracle/src/metabase/driver/oracle.clj
Comment (camsaul) [inline modules/drivers/oracle/src/metabase/driver/oracle.clj:455]: How strong is our conviction that we don't want to be passing `(:schema table)` here?

Comment (camsaul) [reply modules/drivers/oracle/src/metabase/driver/oracle.clj:455]: I know Oracle supports schemas

File: modules/drivers/snowflake/src/metabase/driver/snowflake.clj
Comment (camsaul) [inline modules/drivers/snowflake/src/metabase/driver/snowflake.clj:77]: Is this supposed to be part of this PR?

File: modules/drivers/snowflake/src/metabase/driver/snowflake.clj
Comment (camsaul) [inline modules/drivers/snowflake/src/metabase/driver/snowflake.clj:430]: thanks for adding this

File: src/metabase/driver/sql_jdbc/sync/describe_table.clj
Comment (camsaul) [inline src/metabase/driver/sql_jdbc/sync/describe_table.clj:184]: Don't forget to document driver API changes in `driver-changelog.md`

https://github.com/metabase/metabase/blob/master/docs/developers-guide/driver-changelog.md

**Comment by camsaul** [comment]
What in the world is this PR description. Is every sentence a bullet point?

**Review by camsaul — CHANGES_REQUESTED** [review]
Our policy is to only ship bugfixes and security-related dependency bumps in point releases, and there's a good reason for that.

We absolutely do not want to break anything a point release. There is nothing worse than someone using 45.0 with everything working upgrading to 45.1 and running in to new bugs. Also bad: using 45.0 with one bug bothering them, upgrading to 45.1, which fixes that bug, but running into a new bug. 

Ok, but let's suppose we are 99% sure this change is not going to break anything. We still shouldn't ship it in a point release.

Suppose we decide that any minor feature we are 99% sure will not break anything is perfectly fine to ship in a point release. Thus the chance of any one such change breaking anything is 1%. Suppose we ship an average of 10 such features in point releases for each major release cycle. The change of something breaking is `0.01 * 10`, or 10%. Now suppose we ship a major release every two months — six major releases a year. `0.1 * 6 = 0.6` — the chance of us breaking something in a point release at least once a year is 60%, in other words, more likely than not. I don't like those odds.

If I'm totally off base here, please let me know, and I'll re-review.

Two other notes:
- Please explicitly tag me to review all QP/driver PRs and give me a chance to review them -- I would have liked to review the original PR here. We need to be extra careful with QP stuff and get extra eyes on them
- Don't forget to document driver interface changes

**Comment by escherize** [comment]
Thanks for taking the time to review. I will tag you for sure next time!!

Reading through your comments, I'd say I'm 95% sure this is bug-free. So I'll go ahead and close now.

I'll apply the review comments in a follow-up PR.

**Comment by luizarakaki** [comment]
@camsaul Isn't this a P1 bugfix? This is not a new feature and is closing #26054.

Many customers are reaching out about this bug, costing them ~500 USD/mo for a medium size Snowflake database. This is a legitimate reason to churn. In larger instances, the cost of the primary key sync alone surpasses 1k USD/mo.

https://metabase.zendesk.com/agent/tickets/13469
https://metabase.zendesk.com/agent/tickets/10744
https://metabase.zendesk.com/agent/tickets/12926
and a few others

**Comment by camsaul** [comment]
Oh ok, I was missing that context. It's always a good idea to link the issue in the PR description! I thought this was just a routine performance improvement and not something we were considering to be a P1 performance issue.

In that case I will revoke my request for changes.

**Review by camsaul — COMMENTED** [review]
Well I can't approve it now because it's closed. So tag me again and I will approve once you re-open

**Review by camsaul — APPROVED** [review]

**Pull request merged**
